File: packages/python/src/konfigo/store.py
from __future__ import annotations

from collections.abc import Callable, Iterable, Mapping
from datetime import datetime, timezone
import threading

from konfigo.models import ConfigEntry

TIMESTAMP_KEY = "RealtimeConfigOptions:Timestamp"


class RealtimeConfigStore:
    def __init__(self, entries: Iterable[ConfigEntry] = ()) -> None:
        self._data: dict[str, str | None] = {}
        self._generations: dict[str, int] = {}
        self._timestamp = datetime.min.replace(tzinfo=timezone.utc)
        self._callbacks: list[Callable[[Mapping[str, str | None]], None]] = []
        self._lock = threading.RLock()
        self.update(entries, notify=False)

    @property
    def timestamp(self) -> datetime:
        with self._lock:
            return self._timestamp

    def snapshot(self) -> dict[str, str | None]:
        with self._lock:
            return dict(self._data)

    def get(self, key: str, default: str | None = None) -> str | None:
        with self._lock:
            return self._data.get(key, default)

    def subscribe(self, callback: Callable[[Mapping[str, str | None]], None]) -> Callable[[], None]:
        with self._lock:
            self._callbacks.append(callback)

        def unsubscribe() -> None:
            with self._lock:
                if callback in self._callbacks:
                    self._callbacks.remove(callback)

        return unsubscribe

    def update(self, entries: Iterable[ConfigEntry], *, notify: bool = True) -> bool:
        with self._lock:
            updated = False
            for entry in entries:
                current_generation = self._generations.get(entry.key)
                if current_generation is not None and current_generation >= entry.generation:
                    continue

                self._data[entry.key] = entry.value
                self._generations[entry.key] = entry.generation
                self._timestamp = max(self._timestamp, _normalized(entry.timestamp))
                updated = True

            self._data[TIMESTAMP_KEY] = self._timestamp.isoformat()
            callbacks = tuple(self._callbacks) if updated and notify else ()
            snapshot = dict(self._data) if callbacks else {}

        for callback in callbacks:
            callback(snapshot)

        return updated


def _normalized(value: datetime) -> datetime:
    if value.tzinfo is None:
        return value.replace(tzinfo=timezone.utc)
    return value.astimezone(timezone.utc)

File: packages/go/store.go
package konfigo

import (
	"sync"
	"time"
)

const TimestampKey = "RealtimeConfigOptions:Timestamp"

type Store struct {
	mu          sync.RWMutex
	data        map[string]*string
	generations map[string]int32
	timestamp   time.Time
	callbacks   map[int]func(map[string]*string)
	nextID      int
}

func NewStore(entries ...ConfigEntry) *Store {
	store := &Store{
		data:        map[string]*string{},
		generations: map[string]int32{},
		timestamp:   time.Time{}.UTC(),
		callbacks:   map[int]func(map[string]*string){},
	}
	store.Update(entries, false)
	return store
}

func (s *Store) Timestamp() time.Time {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return s.timestamp
}

func (s *Store) Snapshot() map[string]*string {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return copyData(s.data)
}

func (s *Store) Get(key string) (*string, bool) {
	s.mu.RLock()
	defer s.mu.RUnlock()
	value, ok := s.data[key]
	return cloneString(value), ok
}

func (s *Store) Subscribe(callback func(map[string]*string)) func() {
	s.mu.Lock()
	id := s.nextID
	s.nextID++
	s.callbacks[id] = callback
	s.mu.Unlock()

	return func() {
		s.mu.Lock()
		delete(s.callbacks, id)
		s.mu.Unlock()
	}
}

func (s *Store) Update(entries []ConfigEntry, notify bool) bool {
	s.mu.Lock()

	updated := false
	for _, entry := range entries {
		if current, ok := s.generations[entry.Key]; ok && current >= entry.Generation {
			continue
		}

		s.data[entry.Key] = cloneString(entry.Value)
		s.generations[entry.Key] = entry.Generation
		if ts := normalizeTime(entry.Timestamp); ts.After(s.timestamp) {
			s.timestamp = ts
		}
		updated = true
	}

	timestamp := s.timestamp.Format(time.RFC3339Nano)
	s.data[TimestampKey] = &timestamp

	var callbacks []func(map[string]*string)
	var snapshot map[string]*string
	if updated && notify {
		callbacks = make([]func(map[string]*string), 0, len(s.callbacks))
		for _, callback := range s.callbacks {
			callbacks = append(callbacks, callback)
		}
		snapshot = copyData(s.data)
	}

	s.mu.Unlock()

	for _, callback := range callbacks {
		callback(copyData(snapshot))
	}

	return updated
}

func copyData(data map[string]*string) map[string]*string {
	result := make(map[string]*string, len(data))
	for key, value := range data {
		result[key] = cloneString(value)
	}
	return result
}

func cloneString(value *string) *string {
	if value == nil {
		return nil
	}
	cloned := *value
	return &cloned
}

func normalizeTime(value time.Time) time.Time {
	if value.IsZero() {
		return time.Time{}.UTC()
	}
	return value.UTC()
}
